### PR TITLE
Bundles xapi-db-load into a command-line tool with setuptools

### PR DESCRIPTION
**Pull request by pomegranited**
...and handles a [breaking change with query settings from `clickhouse-connect==0.5.0`](https://github.com/ClickHouse/clickhouse-connect/blob/main/CHANGELOG.md#warning----breaking-change----removing-get_client-arbitrary-keyword-arguments).

**Testing instructions**

E.g inside a docker container with access to `clickhouse_service` from https://github.com/bmtcril/tutor-contrib-clickhouse :

1. Checkout this branch, and change into the repo directory.
1. In a new python virtualenv, run `pip install .` to install it and the latest dependencies.
1. Run `xapi-db-load --backend clickhouse...` to ensure that the script still runs as expected.

**Author Notes & Concerns**

1. @bmtcril Can you tag this as `0.1` after merging? Then I can reference it in my tutor init scripts, and we don't have to worry about api changes on the script.

### CODE REVIEW

**Comment by pomegranited** [comment]
@bmtcril Are you ok with these changes? I had to make them in order to use this code to (optionally) loading demo xAPI data with my OARS tutor plugin.

**Comment by pomegranited** [comment]
@bmtcril and out of interest, do the xAPI schemas used here more closely match Ralph's or the event-routing-backends?

**Comment by bmtcril** [comment]
I'll have to dig through and see what I've got pending. I think this is fine, I just want to make sure any big refactors I've done have already landed otherwise it might be easier to land them first then rebase. 

The events generated here are straight from the event-routing-backends tests right now: https://github.com/openedx/event-routing-backends/tree/master/event_routing_backends/processors/xapi/tests/fixtures/expected

There are definitely some things that can be improved, especially w/r/t ordering of events. Right now I _think_ it enrolls learners in courses before firing other events for them, but everything else happens in a random order and will make for some extremely confusing visualizations.

**Comment by bmtcril** [comment]
Sadly I did have a large branch running for the Ralph tests. If you want to take a spin through it here, I can merge it and then fight through the rebase on this PR to get it in: https://github.com/bmtcril/xapi-db-load/pull/2

**Comment by pomegranited** [comment]
@bmtcril I'm more than happy to rebase -- my changes are simple, they only look big. Please let me know when you've merged your major work, and I'll reapply my changes.

**Comment by bmtcril** [comment]
Ok, I've merged my big mess and will start treating this as a real shared repo now. Thanks!

File: xapi_db_load/backends/clickhouse_lake.py
Comment (pomegranited) [inline xapi_db_load/backends/clickhouse_lake.py:24]: @bmtcril I get this error when I include `'old_parts_lifetime': 10` in the client settings:

> clickhouse_connect.driver.exceptions.ProgrammingError: Setting old_parts_lifetime is unknown or readonly

So I've added it to the MergeTable CREATE statement below instead. From what you've learned about Clickhouse, is that ok?

**Comment by pomegranited** [comment]
Thanks for merging your changes @bmtcril ! I've rebased and re-tested, and fixed some things (see individual commits).
So this PR and https://github.com/bmtcril/tutor-contrib-clickhouse/pull/1 are ready for your review :)

I'm working on a 3rd PR against https://github.com/open-craft/tutor-contrib-superset with scripts to initialize the clickhouse data source, and add a chart + dashboard for https://github.com/openedx/data-wg/issues/26.

File: README.rst
Comment (bmtcril) [inline README.rst:29]: This help is a bit out of date...

```suggestion
    Options:
  --backend [clickhouse|mongo|citus|ralph_clickhouse|ralph_mongo]
                                  Which backend to run against  [required]
  --num_batches INTEGER           Number of batches to run, num_batches *
                                  batch_size is the total rows
  --batch_size INTEGER            Number of rows to insert per batch,
                                  num_batches * batch_size is the total rows
  --drop_tables_first BOOLEAN     If True, the target tables will be dropped
                                  if they already exist
  --distributions_only BOOLEAN    Just run distribution queries and exit
  --db_host TEXT                  Database host name
  --db_port TEXT                  Database port
  --db_name TEXT                  Database name
  --db_username TEXT              Database username
  --db_password TEXT              Password for the database so it's not stored
                                  on disk
  --lrs_url TEXT                  URL to the LRS, if used
  --lrs_username TEXT             LRS username
  --lrs_password TEXT             Password for the LRS so it's not stored on
                                  disk
  --help                          Show this message and exit.
```

File: xapi_db_load/backends/clickhouse_lake.py
Comment (bmtcril) [reply xapi_db_load/backends/clickhouse_lake.py:24]: That should be fine here. I wonder if there's a version mismatch in one of our installs, though. If we're going to be using this elsewhere we should probably do the work to make it an officially supported repo. I'll ask around about what people think about that.

File: xapi_db_load/backends/clickhouse_lake.py
Comment (bmtcril) [inline xapi_db_load/backends/clickhouse_lake.py:52]: These changes are all fine, this is mostly a note for myself to come back and change the rest of this create statement to the one that was performing better in the Ralph tests:

```
            CREATE TABLE IF NOT EXISTS {self.event_table_name} (
            event_id UUID NOT NULL,
            verb_id String NOT NULL,
            actor_id UUID NOT NULL,
            org String NOT NULL,
            course_id String NOT NULL,
            emission_time timestamp NOT NULL,
            event JSON NOT NULL
            )
            ENGINE MergeTree ORDER BY (org, course_id, verb_id, actor_id, emission_time,
             event_id)
            PRIMARY KEY (org, course_id, verb_id, actor_id, emission_time, event_id)
```

**Review by bmtcril — COMMENTED** [review]
This is great, I left some comments but I think the only real change needed is to the help string.

**Comment by bmtcril** [comment]
Just like the other one, @pomegranited this repository has been moved into the openedx GitHub org. Can you confirm here that you're good with contributing this PR to openedx under your CLA?

File: xapi_db_load/backends/clickhouse_lake.py
Comment (pomegranited) [reply xapi_db_load/backends/clickhouse_lake.py:24]: @bmtcril Or maybe we need to pin the version of `clickhouse-connect`? I've got this, what are you running?

```
clickhouse-connect==0.5.3
```

File: README.rst
Comment (pomegranited) [reply README.rst:29]: @bmtcril Hmm.. Since it's probably going to continue to be out of date, I think it's better to remove the USAGE from the README. What do you think?

File: xapi_db_load/backends/clickhouse_lake.py
Comment (pomegranited) [reply xapi_db_load/backends/clickhouse_lake.py:52]: Ah cool -- I've left the extra fields alone (since the data loading requires them), but c106f35b06b414fea1b0450bcb4b9629bcfa6e63 adds those additional fields to `ORDER BY` and `PRIMARY KEY`.

**Comment by pomegranited** [comment]
@bmtcril I've addressed your suggestions, please have a look?

Thanks for moving this repo into the `openedx` fold :)
I'm happy to contribute this PR to openedx under my CLA. Do I need to also request merge access to it as part of my CC role, and get this and the other OARS repos added to the "analytics" group?

File: README.rst
Comment (bmtcril) [reply README.rst:29]: Makes sense, I've added https://github.com/openedx/xapi-db-load/issues/3 to cover writing some more formal docs and a slew of other housekeeping items.

File: xapi_db_load/backends/clickhouse_lake.py
Comment (bmtcril) [reply xapi_db_load/backends/clickhouse_lake.py:24]: I've got 0.3.6 so am on the other side of this breaking change: https://github.com/ClickHouse/clickhouse-connect/blob/main/CHANGELOG.md#050-2023-01-14

I think pinning it to >0.5 and <0.6 is probably the best since they seem to do breaking changes on minor versions.

File: xapi_db_load/backends/clickhouse_lake.py
Comment (bmtcril) [reply xapi_db_load/backends/clickhouse_lake.py:52]: Ok, thanks. I'll normalize those between the clickhouse and Ralph backends in a different PR. That table structure is going to all need to be updated after some discussions I just had with FUN, but I haven't decided on exactly how yet.

**Review by bmtcril — DISMISSED** [review]
I'm happy to approve this as-is and resolve the outstanding comments in a follow-on PR so as to unblock https://github.com/openedx/tutor-contrib-clickhouse/pull/1

**Comment by bmtcril** [comment]
@pomegranited I added you as a grandfathered-in "writer" to this repo and tutor-contrib-clickhouse since you were already working on them. I'm happy to sponsor you to the analytics CC group, which would likely make all the management easier. These repos are all in there as well, if you're ok with it I'll make a post about it in the forums!

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by pomegranited** [comment]
@bmtcril 

> I added you as a grandfathered-in "writer" to this repo and tutor-contrib-clickhouse since you were already working on them.

Thank you :) I'll merge this, and get it tagged.

> I'm happy to sponsor you to the analytics CC group, which would likely make all the management easier. These repos are all in there as well, if you're ok with it I'll make a post about it in the forums!

Thank you, yes please! I'm already a [core contributor to the legacy analytics repos](https://discuss.openedx.org/t/coding-cc-rights-expansion-jillian-vogel/7339), but this was granted before we had these CC repo groups.

**Comment by pomegranited** [comment]
@bmtcril Ach, sorry.. I squashed commits and git wants a new approving review. (Tried reverting back to the previously approved commit, and it didn't buy it.) I'll remember for next time and just squash on merge!

**Review by bmtcril — APPROVED** [review]

**Comment by openedx-webhooks** [comment]
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/xapi-db-load&entry.867055334=https://github.com/openedx/xapi-db-load/pull/1&entry.1484655318=https://github.com/pomegranited&entry.752974735=2023-01-25+10:41&entry.1917517419=2023-02-06+13:57&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
